### ISSUE TITLE
Is ActivityRegularization broken?

### ISSUE THREAD

**Issue by RaffEdwardBAH** [fchollet/keras#2767]
I'm trying to implement a custom regularization over the activation of a hidden layer, so I copied the code from `ActivityRegularization` as a starting point. However, I'm not sure `ActivityRegularization` does anything at all. It seems like the ``__call__`` method is never being called (I added an infinite loop yet training progressed) 

If I add a simple layer into a network, like this  `x = ActivityRegularization(l2=1000.0)(x)` , the model learns exactly as it did without any regularization added to the network at all. I'm using the head version of Keras, and have tried it with such absurdity like `ActivityRegularization(l2=100000.0, l1=100000.0)` and observe no impact. Surely this is enough regularization that just learning to output zero is the best strategy?

**Comment by fchollet** [comment]
I don't see any code to reproduce your issue?

**Comment by RaffEdwardBAH** [comment]
It can be easily re-created, just did it by modifying the mnist example. 

```python
'''Trains a simple deep NN on the MNIST dataset.

Gets to 98.40% test accuracy after 20 epochs
(there is *a lot* of margin for parameter tuning).
2 seconds per epoch on a K520 GPU.
'''

from __future__ import print_function
import numpy as np
np.random.seed(1337)  # for reproducibility

from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, ActivityRegularization
from keras.optimizers import SGD, Adam, RMSprop
from keras.utils import np_utils


batch_size = 128
nb_classes = 10
nb_epoch = 20

# the data, shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.reshape(60000, 784)
X_test = X_test.reshape(10000, 784)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)

model = Sequential()
model.add(Dense(512, input_shape=(784,)))
model.add(Activation('relu'))
model.add(Dropout(0.2))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(ActivityRegularization(l1=100000.0, l2=100000.0))
model.add(Dropout(0.2))
model.add(Dense(10))
model.add(Activation('softmax'))

model.summary()

model.compile(loss='categorical_crossentropy',
              optimizer=RMSprop(),
              metrics=['accuracy'])

history = model.fit(X_train, Y_train,
                    batch_size=batch_size, nb_epoch=nb_epoch,
                    verbose=1, validation_data=(X_test, Y_test))
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])
```

Which has the output:

```
Using Theano backend.
Downloading data from https://s3.amazonaws.com/img-datasets/mnist.pkl.gz
15302656/15296311 [==============================] - 2s     
60000 train samples
10000 test samples
____________________________________________________________________________________________________
Layer (type)                       Output Shape        Param #     Connected to                     
====================================================================================================
dense_1 (Dense)                    (None, 512)         401920      dense_input_1[0][0]              
____________________________________________________________________________________________________
activation_1 (Activation)          (None, 512)         0           dense_1[0][0]                    
____________________________________________________________________________________________________
dropout_1 (Dropout)                (None, 512)         0           activation_1[0][0]               
____________________________________________________________________________________________________
dense_2 (Dense)                    (None, 512)         262656      dropout_1[0][0]                  
____________________________________________________________________________________________________
activation_2 (Activation)          (None, 512)         0           dense_2[0][0]                    
____________________________________________________________________________________________________
activityregularization_1 (ActivityR(None, 512)         0           activation_2[0][0]               
____________________________________________________________________________________________________
dropout_2 (Dropout)                (None, 512)         0           activityregularization_1[0][0]   
____________________________________________________________________________________________________
dense_3 (Dense)                    (None, 10)          5130        dropout_2[0][0]                  
____________________________________________________________________________________________________
activation_3 (Activation)          (None, 10)          0           dense_3[0][0]                    
====================================================================================================
Total params: 669706
____________________________________________________________________________________________________
Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 14s - loss: 0.2452 - acc: 0.9246 - val_loss: 0.1033 - val_acc: 0.9683
Epoch 2/20
60000/60000 [==============================] - 14s - loss: 0.1024 - acc: 0.9694 - val_loss: 0.1009 - val_acc: 0.9671
Epoch 3/20
60000/60000 [==============================] - 15s - loss: 0.0751 - acc: 0.9771 - val_loss: 0.0859 - val_acc: 0.9750
Epoch 4/20
60000/60000 [==============================] - 16s - loss: 0.0606 - acc: 0.9815 - val_loss: 0.0743 - val_acc: 0.9779
Epoch 5/20
60000/60000 [==============================] - 19s - loss: 0.0508 - acc: 0.9845 - val_loss: 0.0720 - val_acc: 0.9815
Epoch 6/20
60000/60000 [==============================] - 18s - loss: 0.0445 - acc: 0.9871 - val_loss: 0.0769 - val_acc: 0.9811
Epoch 7/20
60000/60000 [==============================] - 18s - loss: 0.0369 - acc: 0.9884 - val_loss: 0.0828 - val_acc: 0.9809
Epoch 8/20
60000/60000 [==============================] - 18s - loss: 0.0330 - acc: 0.9898 - val_loss: 0.0884 - val_acc: 0.9794
Epoch 9/20
60000/60000 [==============================] - 18s - loss: 0.0310 - acc: 0.9913 - val_loss: 0.0911 - val_acc: 0.9820
Epoch 10/20
60000/60000 [==============================] - 18s - loss: 0.0293 - acc: 0.9914 - val_loss: 0.0873 - val_acc: 0.9830
```

Yet we would expect the accuracy to be terrible, because such an absurdly strong activity regularization should have forced it to learn to always produces zeros for its activations.

**Comment by fchollet** [comment]
The regularizer wasn't being registered. This is fixed: https://github.com/fchollet/keras/commit/24501d4361606d6305e9eaeadaf1cd1bbb8dfab6
